src/coreclr/jit/emit.cpp

```diff
@@ -4479,6 +4610,429 @@ void emitter::emitJumpDistBind()
+            }
+            else
+            {
+                unsigned paddingToAdj = actualPaddingNeeded;
+
+#ifdef DEBUG
+
+                int instrAdjusted = (alignmentBoundary + 14) / 15;
+#endif
+                // Adjust the padding amount in all align instructions in this IG
+                instrDescAlign *alignInstrToAdj = alignInstr, *prevAlignInstr = nullptr;
+                for (; alignInstrToAdj != nullptr && alignInstrToAdj->idaIG == alignInstr->idaIG;
+                     alignInstrToAdj = alignInstrToAdj->idaNext)
+                {
+                    unsigned newPadding = min(paddingToAdj, 15);
+                    alignInstrToAdj->idCodeSize(newPadding);
+                    paddingToAdj -= newPadding;
+                    prevAlignInstr = alignInstrToAdj;
+#ifdef DEBUG
+                    instrAdjusted--;
+#endif
+                }
+                assert(paddingToAdj == 0);
+                assert(instrAdjusted == 0);
+
+                // fast forward the align instruction to next IG
+                alignInstr = prevAlignInstr;
+            }
+
+            JITDUMP("Adjusted alignment of G_M%03u_IG%02u from %02d to %02d\n", emitComp->compMethodID, alignIG->igNum,
+                    estimatedPaddingNeeded, actualPaddingNeeded);
+        }
+
+        if (actualPaddingNeeded > 0)
+        {
+            // Record the last IG that has align instruction. No overestimation
+            // adjustment will be done after emitLastAlignedIgNum.
+            emitLastAlignedIgNum = alignIG->igNum;
+        }
+    }
+
+    // Do adjustments of remaining IGs
+    while (lastIGAdj != nullptr)
+    {
+        lastIGAdj->igOffs -= alignBytesRemoved;
+        lastIGAdj = lastIGAdj->igNext;
+    }
+
+#ifdef DEBUG
+    emitCheckIGoffsets();
+#endif
+}
+
+//-----------------------------------------------------------------------------
+//  emitCalculatePaddingForLoopAlignment: Calculate the padding to insert at the
+//    end of 'ig' so the loop that starts after 'ig' is aligned.
+//
+//  Returns: Padding amount.
+//    0 means no padding is needed, either because loop is already aligned or it
+//    is too expensive to align loop and hence it will not be aligned.
+//
+//  Notes:
+//     Below are the steps (in this order) to calculate the padding amount.
+//     1. If loop is already aligned to desired boundary, then return 0. // already aligned
+//     2. If loop size exceed maximum allowed loop size, then return 0.  // already aligned
+//
+// For adaptive loop alignment:
+//     3a. Calculate paddingNeeded and maxPaddingAmount to align to 32B boundary.
+//     3b. If paddingNeeded > maxPaddingAmount, then recalculate to align to 16B boundary.
+//     3b. If paddingNeeded == 0, then return 0. // already aligned at 16B
+//     3c. If paddingNeeded > maxPaddingAmount, then return 0. // expensive to align
+//     3d. If the loop already fits in minimum 32B blocks, then return 0. // already best aligned
```

Comment:
Why wouldn't we check this first? If padding can't reduce the number of chunks then it seems like we should not pad..?

Reply:
I first try to check the best alignment boundary as part of 3b. Only if we decide to move ahead, we will check for the 3d. So in 3d, i check the current offset of loop with regards to alignmentBoundary (`32B or `16B`). If we do it before, we might be just checking the current offset % 32.